app/src/features/rules/screens/rulesList/components/RulesList/hooks/useFetchAndUpdateRules.tsx

```diff
@@ -43,7 +53,38 @@ const useFetchAndUpdateRules = ({ setIsLoading }: Props) => {
     Promise.all([groupsPromise, rulesPromise])
       .then(async (data) => {
         const groups = data[0] as Group[];
-        const rules = data[1] as Rule[];
+        let rules = data[1] as Rule[];
+
+        try {
+          if (isExtensionManifestVersion3()) {
+            const currentWorkspaceId = activeWorkspace?.id ?? "private";
+            const mv3MigrationData = getMV3MigrationData();
+
+            if (!mv3MigrationData?.[currentWorkspaceId]?.rulesMigrated) {
+              //TODO: rules types are not standardized. Need to standardize them
+
+              //@ts-ignore
+              const rulesMigrationData = detectAndGenerateMV3RulesMigrationData(rules, currentWorkspaceId);
+
+              //@ts-ignore
+              rules = rules.map((rule) => migrateRuleToMV3(rule));
+
+              StorageService(appMode)
+                .saveMultipleRulesOrGroups(rules, { workspaceId: activeWorkspace.id })
+                .then(() => {
+                  saveMV3MigrationData({
+                    ...mv3MigrationData,
+                    [currentWorkspaceId]: {
+                      rulesMigrated: true,
+                      rulesMigrationData: rulesMigrationData,
+                    },
+                  });
+                });
+            }
```

Comment:
This all should happen in a single method. Check [this](https://github.com/requestly/requestly/pull/1669#discussion_r1595136981) comment